### PR TITLE
Rewrite file_script from python to ts. Add variant that consults binary cache.

### PR DESCRIPTION
**Pull request by ras0219-msft**
Today, the files list does not reliably contain the changed files in PR builds. It's calculated based on the final install tree at the end of the `ci` command, which avoids installing packages that aren't depended upon by builds and aren't built. Namely, if the packages were built by feature testing (a previous step) or by a previous PR build, they won't be present.

This PR solves that problem by retrieving the files from the binary cache directly. We consult git for the modified files, the `--output-hashes` for the ABIs, and then download the zips directly for interrogation.

The previous script is still appropriate for CI runs, which always do full builds. However, it was rewritten in typescript to reduce the number of languages we depend upon (javascript was already inherent to AzDO Pipelines / GHA, we use it for the website, etc).

### CODE REVIEW

File: scripts/azure-pipelines/linux/azure-pipelines.yml [hunk 1]
```diff
@@ -102,8 +104,20 @@ jobs:
       targetPath: '$(WORKING_ROOT)/failure-logs'
       artifact: "failure logs for ${{ replace(parameters.jobName, '_', '-') }}"
     condition: ne(variables['FAILURE_LOGS_EMPTY'], 'True')
+  - task: UseNode@1
+    displayName: 'Ensure Node.js is available'
+    inputs:
+      version: '22.x'
   - bash: |
-      python3 scripts/file_script.py /mnt/vcpkg-ci/installed/vcpkg/info/
+      cd scripts/azure-pipelines/owners-db && npm ci || true
```
Comment (BillyONeal) [inline scripts/azure-pipelines/linux/azure-pipelines.yml:112]: Can we move this into the AzureCLI task above so that we don't have to persist secret material in a file?

**Comment by Thomas1664** [comment]
Great timing for rewriting a perfectly fine Python script with 0 dependencies to typescript with 15+ dependencies after this weeks supply chain attack.
As far as I understand the code, the new functionality of downloading artifacts from the cache can also be done with 0 dependencies in Python.
By reducing the number of dependencies you not only decrease the risk of supply chain attacks but also avoid the additional maintanace work of keeping your Nodejs dependencies up to date.

**Comment by BillyONeal** [comment]
The problem is that the python dependency itself is problematic, while we are already using Azure DevOps and GitHub Actions, both of which are creatures of NodeJS.

File: scripts/azure-pipelines/linux/azure-pipelines.yml (hunk 1 above)
Comment (ras0219-msft) [reply scripts/azure-pipelines/linux/azure-pipelines.yml:112]: The tricky thing is that we want to run the file lists script even when the main vcpkg process fails; theoretically we could do this in each shell, but it seems more "obviously correct" to use `condition: always()`

**Comment by ras0219-msft** [comment]
Back to draft to solve:
```
(node:6371) [MODULE_TYPELESS_PACKAGE_JSON] Warning: Module type of file:///mnt/vss/_work/1/s/scripts/azure-pipelines/owners-db/file_script_from_cache.ts is not specified and it doesn't parse as CommonJS.
Reparsing as ES module because module syntax was detected. This incurs a performance overhead.
To eliminate this warning, add "type": "module" to /mnt/vss/_work/1/s/scripts/azure-pipelines/owners-db/package.json.
(Use `node --trace-warnings ...` to show where the warning was created)
```
and
```
Downloading https://vcpkgbinarycachewus.blob.core.windows.net/cache/f6598db5cadac7a84667ae3b75f6e1a82a32447b9a78b08e0be8174995dc8829.zip for port wg21-linear-algebra...
Failed to download or process blob for wg21-linear-algebra: Error: HTTP 409 while fetching https://vcpkgbinarycachewus.blob.core.windows.net/cache/f6598db5cadac7a84667ae3b75f6e1a82a32447b9a78b08e0be8174995dc8829.zip
```

**Comment by ras0219-msft** [comment]
Ok, both issues appear solved:
- See https://github.com/microsoft/vcpkg/pull/47324 / https://dev.azure.com/vcpkg/public/_build/results?buildId=120294&view=logs&jobId=ac0731cb-f2f9-5764-2e0a-a6fa8ebb4a08&j=18f767b8-9e55-53aa-9c57-42862973482f&t=b58e43b3-b77e-5af7-6e46-835c7297d3e7 for an example of the secret successfully used and redacted
- See this PR for the commonjs modules fix

I've tested the CI variant (file_script.ts) locally and confirmed it works as advertised as well; what I can't fully test in PR is whether the `info-dir` parameters are correct on all platforms. After merging, the active infra owner should check the next CI results to ensure things are working as expected.

File: scripts/azure-pipelines/owners-db/file_script.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    const components = file.split("_");
+    const pkg = components[0] + ":" + components[2].replace(".list", "");
+    const content = fs.readFileSync(path.join(dirPath, file), "utf8");
+    const lines = content.split(/\r?\n/);
+    for (const raw of lines) {
+      if (!raw) continue;
+      const line = raw.trim();
+      if (line.length === 0) continue;
+      if (line.endsWith("/")) continue;
+      const idx = line.indexOf("/");
```
Comment (BillyONeal) [inline scripts/azure-pipelines/owners-db/file_script.ts:29]:
```suggestion
      # remove the leading triplet directory name
      const idx = line.indexOf("/");
```

File: scripts/azure-pipelines/owners-db/file_script.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    process.exit(2);
+  }
+  return { infoDir, outDir };
+}
+
+function main() {
+  const { infoDir: dir, outDir } = parseArgs(process.argv.slice(2));
+  try {
+    fs.mkdirSync(outDir, { recursive: true });
+  } catch (err) {
```
Comment (BillyONeal) [inline scripts/azure-pipelines/owners-db/file_script.ts]:
```suggestion
  } catch {
```

File: scripts/azure-pipelines/owners-db/file_script.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env node
+import * as fs from "fs";
+import * as path from "path";
+
+const keyword = "/include/";
```
Comment (BillyONeal) [inline scripts/azure-pipelines/owners-db/file_script.ts]:
```suggestion
const include_path = "/include/";
```

keyword is awfully confusing

File: scripts/azure-pipelines/owners-db/file_script_from_cache.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+function usage() {
+  console.error("Usage: file_script_from_cache.ts --pr-hashes <pr-hashes.json> --blob-base-url <blob-base-url> [--target-branch <branch>] [--out-dir <path>]");
+  console.error("blob-base-url should include SAS token (e.g. https://<account>.blob.core.windows.net/<container>/?<sas>)");
+}
+
+function parseArgs(argv: string[]) {
+  // supports: --pr-hashes <path> --blob-base-url <url> [--target-branch <branch>] [--out-dir <path>]
+  // legacy: positional: <pr-hashes> <blob-base-url> [target-branch]
```
Comment (BillyONeal) [inline scripts/azure-pipelines/owners-db/file_script_from_cache.ts:51]: Do we really need a legacy interface for our own support script like this?

File: scripts/azure-pipelines/owners-db/file_script_from_cache.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  const dbLines: string[] = [];
+  const headerLines: string[] = [];
+
+  // Determine list of ports to process from git-diff (only folders under ports/ that changed)
+  let changedPorts: string[] = [];
+  try {
+    const gitRange = `${targetBranch}...HEAD`;
+    const diffOut = execSync(`cd ../../.. && git diff --name-only ${gitRange} -- ports/`, { encoding: "utf8" });
```
Comment (BillyONeal) [inline scripts/azure-pipelines/owners-db/file_script_from_cache.ts]: Where does CD need to be initially for the ../../.. to be correct? (It just seems a strange hard coding in this spot where everything else here appears to be trying to avoid such assumptions)

File: scripts/azure-pipelines/test-modified-ports.ps1 [hunk 7]
```diff
@@ -229,7 +229,19 @@ if ($AllowUnexpectedPassing) {
 Add-ToolchainToTestCMake
 $xunitFile = Join-Path $ArtifactStagingDirectory "$Triplet-results.xml"
 $xunitArg = "--x-xunit=$xunitFile"
-& $vcpkgExe ci $tripletArg $failureLogsArg $xunitArg $ciBaselineArg @commonArgs @cachingArgs @parentHashesArgs @skipFailuresArgs @knownFailuresFromArgs @allowUnexpectedPassingArgs
+$prHashes = Join-Path $ArtifactStagingDirectory "pr-hashes.json"
```
Comment (BillyONeal) [inline scripts/azure-pipelines/test-modified-ports.ps1]:
```suggestion
$prHashesFile = Join-Path $ArtifactStagingDirectory "pr-hashes.json"
```

Note the convention elsewhere of fooFile for the file path and fooArg for fooFile with the --arg attached.

File: scripts/azure-pipelines/owners-db/package.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "owners-db",
+  "version": "1.0.0",
+  "description": "Utility to generate VCPKG file lists (TypeScript)",
+  "private": true,
+  "type": "module",
+  "license": "MIT",
+  "engines": {
+    "node": ">=14"
```
Comment (BillyONeal) [inline scripts/azure-pipelines/owners-db/package.json]:
```suggestion
    "node": ">=22"
```

File: scripts/azure-pipelines/owners-db/file_script_from_cache.ts [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    process.exit(2);
+  }
+
+  const dbLines: string[] = [];
+  const headerLines: string[] = [];
+
+  // Determine list of ports to process from git-diff (only folders under ports/ that changed)
+  let changedPorts: string[] = [];
+  try {
+    const gitRange = `${targetBranch}...HEAD`;
```
Comment (BillyONeal) [inline scripts/azure-pipelines/owners-db/file_script_from_cache.ts]: I believe this diff is incorrect. We aren't looking at the ports difference between the target branch and HEAD, we are looking at the ports difference between the *merge base* and HEAD.

Maybe we should add a `z-` command that prints the same determination `x-test-features --for-merge-with` uses because it handles such edge cases.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
I think the HEAD/master diff is incorrect. Everything else is just nitpicks.

**Review by BillyONeal — APPROVED** [review]
